lint: add array range copy checker

For range loops of form `for k, v := range a`, where `a` is an array,
emit a warning if array is big enough to cause measurable impact on performance due to copying.

This code triggers a warning:
```go
var xs [256]byte
/// copy of xs (256 bytes) can be avoided with &xs
for _, x := range xs {
	_ = x
}
```

There are corner cases that must be handled.
For example, if range expression is not addressable, we can't suggest `&a` alternative:
```go
func returnArray() [20]int {
	return [20]int{}
}

func noWarnings() {
	// OK: returned valus is not addressible, can't take address.
	for _, x := range returnArray() {
		_ = x
	}
}
```

The size threshold is needed to avoid bad advices.
Changing small array loop that copies data to `&a` form can cause slowdown on some
machines. Empirically found more-or-less optimal threshold of 64 bytes (6 machine words on x64), but we should probably stick to 96 bytes to avoid neutral changes that give no stable benefits.

Checker suggested by @fexolm.

Running it over GOROOT with lintwalk:
```
$GOROOT/src/cmd/compile/internal/ssa/compile.go:54:2: array-range: copy of passes (2816 bytes) can be avoided with &passes
$GOROOT/src/cmd/compile/internal/ssa/compile.go:196:3: array-range: copy of passes (2816 bytes) can be avoided with &passes
$GOROOT/src/cmd/compile/internal/ssa/compile.go:313:2: array-range: copy of passes (2816 bytes) can be avoided with &passes
$GOROOT/src/cmd/compile/internal/ssa/compile.go:469:2: array-range: copy of passOrder (832 bytes) can be avoided with &passOrder
$GOROOT/src/cmd/compile/internal/ssa/compile.go:473:3: array-range: copy of passes (2816 bytes) can be avoided with &passes
$GOROOT/src/cmd/internal/obj/link.go:495:2: array-range: copy of textAttrStrings (336 bytes) can be avoided with &textAttrStrings
$GOROOT/src/cmd/compile/internal/gc/main.go:916:2: array-range: copy of runtimeDecls (4512 bytes) can be avoided with &runtimeDecls
$GOROOT/src/cmd/compile/internal/gc/main.go:1243:2: array-range: copy of Debug (2048 bytes) can be avoided with &Debug
$GOROOT/src/cmd/compile/internal/gc/ssa.go:2775:2: array-range: copy of sys.Archs (104 bytes) can be avoided with &sys.Archs
$GOROOT/src/cmd/compile/internal/gc/universe.go:71:2: array-range: copy of builtinFuncs (353 bytes) can be avoided with &builtinFuncs
$GOROOT/src/cmd/compile/internal/gc/universe.go:97:2: array-range: copy of basicTypes (329 bytes) can be avoided with &basicTypes
$GOROOT/src/cmd/compile/internal/gc/universe.go:116:2: array-range: copy of builtinFuncs (353 bytes) can be avoided with &builtinFuncs
$GOROOT/src/cmd/internal/objfile/goobj.go:116:2: array-range: copy of sys.Archs (104 bytes) can be avoided with &sys.Archs
$GOROOT/src/crypto/elliptic/p256_amd64.go:340:2: array-range: copy of p.xyz (96 bytes) can be avoided with &p.xyz
$GOROOT/src/go/types/universe.go:99:2: array-range: copy of predeclaredConsts (120 bytes) can be avoided with &predeclaredConsts
$GOROOT/src/strings/replace.go:259:2: array-range: copy of r.mapping (256 bytes) can be avoided with &r.mapping
$GOROOT/src/strings/replace.go:264:2: array-range: copy of r.mapping (256 bytes) can be avoided with &r.mapping
$GOROOT/src/image/jpeg/reader.go:726:2: array-range: copy of translations (128 bytes) can be avoided with &translations
$GOROOT/src/image/jpeg/writer.go:205:2: array-range: copy of theHuffmanSpec (160 bytes) can be avoided with &theHuffmanSpec
$GOROOT/src/runtime/mprof.go:471:2: array-range: copy of r.Stack0 (256 bytes) can be avoided with &r.Stack0
$GOROOT/src/runtime/mprof.go:514:2: array-range: copy of r.Stack0 (256 bytes) can be avoided with &r.Stack0
$GOROOT/src/runtime/trace.go:843:2: array-range: copy of tab.tab (65536 bytes) can be avoided with &tab.tab
$GOROOT/src/runtime/traceback.go:624:2: array-range: copy of cgoPCs (256 bytes) can be avoided with &cgoPCs
$GOROOT/src/compress/flate/inflate.go:235:3: array-range: copy of h.chunks (2048 bytes) can be avoided with &h.chunks
```

--- a/cmd/kfulint/testdata/range-expr-copy/checker_tests.go
+++ b/cmd/kfulint/testdata/range-expr-copy/checker_tests.go
@@ -1,0 +1,63 @@
+package checker_test
+
+func warnings() {
+	{
+		var xs [256]byte
+		/// copy of xs (256 bytes) can be avoided with &xs
+		for _, x := range xs {
+			_ = x
+		}
+	}
+
+	{
+		var foo struct {
+			arr [400]byte
+		}
+		/// copy of foo.arr (400 bytes) can be avoided with &foo.arr
+		for _, x := range foo.arr {
+			_ = x
+		}
+	}
+
+	{
+		xsList := make([][512]byte, 1)
+		/// copy of xsList[0] (512 bytes) can be avoided with &xsList[0]
+		for _, x := range xsList[0] {
+			_ = x
+		}
+	}
+}
+
+func returnArray() [20]int {
+	return [20]int{}
+}
+
+func noWarnings() {
+	// OK: returned valus is not addressible, can't take address.
+	for _, x := range returnArray() {
+		_ = x
+	}
+
+	{
+		var xs [200]byte
+		// OK: already iterating over a pointer.
+		for _, x := range &xs {
+			_ = x
+		}
+		// OK: only index is used. No copy is generated.
+		for i := range xs {
+			_ = xs[i]
+		}
+		// OK: like in case above, no copy, so it's OK.
+		for range xs {
+		}
+	}
+
+	{
+		var xs [10]byte
+		// OK: xs is a very small array that can be trivially copied.
+		for _, x := range xs {
+			_ = x
+		}
+	}
+}

--- a/lint/big-copy_checker.go
+++ b/lint/big-copy_checker.go
@@ -2,8 +2,6 @@ package lint
 
 import (
 	"go/ast"
-	"go/types"
-	"strings"
 )
 
 // bigCopyCheck finds places where big value copy could be unexpected.
@@ -21,43 +19,24 @@ type bigCopyChecker struct {
 }
 
 func (c *bigCopyChecker) PerFuncInit(fn *ast.FuncDecl) bool {
-	return fn.Body != nil && !c.isUnitTestFunc(fn)
+	return fn.Body != nil && !c.ctx.IsUnitTestFuncDecl(fn)
 }
 
 func (c *bigCopyChecker) CheckStmt(stmt ast.Stmt) {
-	if rng, ok := stmt.(*ast.RangeStmt); ok {
-		c.checkRangeStmt(rng)
-	}
-}
-
-func (c *bigCopyChecker) checkRangeStmt(rng *ast.RangeStmt) {
-	if rng.Value == nil {
+	rng, ok := stmt.(*ast.RangeStmt)
+	if !ok || rng.Value == nil {
 		return
 	}
-	const sizeThreshold = 48
 	typ := c.ctx.TypesInfo.TypeOf(rng.Value)
 	if typ == nil {
 		return
 	}
-	size := c.ctx.SizesInfo.Sizeof(typ)
-	if size > sizeThreshold {
-		c.warnRangeValue(rng, size)
+	const sizeThreshold = 48
+	if size := c.ctx.SizesInfo.Sizeof(typ); size >= sizeThreshold {
+		c.warn(rng, size)
 	}
 }
 
-// isUnitTestFunc reports whether FuncDecl declares testing function.
-func (c *bigCopyChecker) isUnitTestFunc(fn *ast.FuncDecl) bool {
-	if !strings.HasPrefix(fn.Name.Name, "Test") {
-		return false
-	}
-	typ := c.ctx.TypesInfo.TypeOf(fn.Name)
-	if sig, ok := typ.(*types.Signature); ok {
-		return sig.Results().Len() == 0 &&
-			sig.Params().Len() == 1
-	}
-	return false
-}
-
-func (c *bigCopyChecker) warnRangeValue(node ast.Node, size int64) {
+func (c *bigCopyChecker) warn(node ast.Node, size int64) {
 	c.ctx.Warn(node, "each iteration copies %d bytes (consider pointers or indexing)", size)
 }

--- a/lint/lint.go
+++ b/lint/lint.go
@@ -30,6 +30,7 @@ var checkFunctions = map[string]struct {
 	"comments":          {new: commentsCheck},
 	"unexported-call":   {new: unexportedCallCheck},
 	"builtin-shadow":    {new: builtinShadowCheck},
+	"range-expr-copy":   {new: rangeExprCopyCheck},
 }
 
 // RuleList returns a slice of all rules that can be used to create checkers.

--- a/lint/range-expr-copy_checker.go
+++ b/lint/range-expr-copy_checker.go
@@ -1,0 +1,52 @@
+package lint
+
+import (
+	"go/ast"
+	"go/types"
+)
+
+// rangeExprCopyCheck detects for statements with range expressions
+// that perform excessive copying (big arrays can cause it).
+//
+// Suggested fix is to use & and doing a range over pointer to array.
+//
+// Rationale: performance.
+func rangeExprCopyCheck(ctx *context) func(*ast.File) {
+	// TODO(quasilyte): there is some annoying code duplication with other
+	// range statement checker. We should consider refactoring if
+	// more checkers that inspect range statements will appear.
+	return wrapStmtChecker(&rangeExprCopyChecker{
+		baseStmtChecker: baseStmtChecker{ctx: ctx},
+	})
+}
+
+type rangeExprCopyChecker struct {
+	baseStmtChecker
+}
+
+func (c *rangeExprCopyChecker) PerFuncInit(fn *ast.FuncDecl) bool {
+	return fn.Body != nil && !c.ctx.IsUnitTestFuncDecl(fn)
+}
+
+func (c *rangeExprCopyChecker) CheckStmt(stmt ast.Stmt) {
+	rng, ok := stmt.(*ast.RangeStmt)
+	if !ok || rng.Key == nil || rng.Value == nil {
+		return
+	}
+	tv := c.ctx.TypesInfo.Types[rng.X]
+	if !tv.Addressable() {
+		return
+	}
+	if _, ok := tv.Type.(*types.Array); !ok {
+		return
+	}
+	const sizeThreshold = 96 // Not recommended to set value lower than 64
+	if size := c.ctx.SizesInfo.Sizeof(tv.Type); size >= sizeThreshold {
+		c.warn(rng, size)
+	}
+}
+
+func (c *rangeExprCopyChecker) warn(rng *ast.RangeStmt, size int64) {
+	x := nodeString(c.ctx.FileSet, rng.X)
+	c.ctx.Warn(rng, "copy of %s (%d bytes) can be avoided with &%s", x, size, x)
+}

--- a/lint/util.go
+++ b/lint/util.go
@@ -5,6 +5,8 @@ import (
 	"go/ast"
 	"go/printer"
 	"go/token"
+	"go/types"
+	"strings"
 )
 
 // unparen returns expr without surrounding parenthesis.
@@ -24,4 +26,18 @@ func nodeString(fset *token.FileSet, x ast.Node) string {
 		panic(err)
 	}
 	return buf.String()
+}
+
+// IsUnitTestFunc reports whether FuncDecl declares testing function.
+func (ctx *context) IsUnitTestFuncDecl(fn *ast.FuncDecl) bool {
+	if !strings.HasPrefix(fn.Name.Name, "Test") {
+		return false
+	}
+	typ := ctx.TypesInfo.TypeOf(fn.Name)
+	if sig, ok := typ.(*types.Signature); ok {
+		return sig.Results().Len() == 0 &&
+			sig.Params().Len() == 1 &&
+			sig.Params().At(0).Type().String() == "*testing.T"
+	}
+	return false
 }

